packages/shared/src/SliderLabels.js

```diff
@@ -0,0 +1,177 @@
+// @flow
+
+import * as React from 'react';
+import { defaultTokens } from '@kiwicom/mobile-orbit';
+import { View } from 'react-native';
+import type { OnLayout } from '@kiwicom/mobile-shared';
+import type { TranslationType } from '@kiwicom/mobile-localization';
+
+import Text from './Text';
+import StyleSheet from './PlatformStyleSheet';
+
+type Props = {|
+  +start: {|
+    +label: TranslationType,
+    +value: number,
+  |},
+  +end?: {|
+    +label: TranslationType,
+    +value: number,
+  |},
+  +max: number,
+  +min: number,
+|};
+
+type State = {|
+  width: number,
+  labelStartWidth: number,
+  labelStartAtMax: boolean,
+  labelEndWidth: number,
+  paddingLeft: number,
+  paddingRight: number,
+|};
+
+export default class SliderLabels extends React.Component<Props, State> {
+  constructor() {
+    super();
+    this.state = {
+      width: 0,
+      labelStartWidth: 0,
+      labelStartAtMax: false,
+      labelEndWidth: 0,
+      paddingLeft: 0,
+      paddingRight: 0,
+    };
+  }
+
+  componentDidMount = () => {
+    requestAnimationFrame(this.setPadding);
+  };
+
+  getMaxPadding = () => {
+    const gap = 10;
+
+    if (this.props.end) {
+      return Math.floor(
+        this.state.width -
+          this.state.labelStartWidth -
+          this.state.labelEndWidth -
+          gap,
+      );
+    } else {
+      return Math.floor(this.state.width - this.state.labelStartWidth - gap);
+    }
+  };
+
+  calculateMarkerOffset = (type: 'start' | 'end'): number => {
+    const { min, max, start, end } = this.props;
+
+    if (type === 'start') {
+      return Math.round((start.value / (max - min)) * this.state.width);
+    } else if (type === 'end' && end) {
+      const val = (end.value / (max - min)) * this.state.width;
+      return Math.round(this.state.width - val);
+    } else {
+      return 0;
+    }
+  };
+
+  setPadding = () => {
+    const startMarkerOffset = this.calculateMarkerOffset('start');
+    const maxPadding = this.getMaxPadding();
+
+    const startLabelHalf = this.state.labelStartWidth / 2;
+    const startLabelOffset =
+      startMarkerOffset < startLabelHalf
+        ? 0
+        : startMarkerOffset - startLabelHalf;
+
+    if (this.props.end) {
+      const endMarkerOffset = this.calculateMarkerOffset('end');
+      const endLabelHalf = this.state.labelEndWidth / 2;
+      const endLabelOffset =
+        endMarkerOffset < endLabelHalf ? 0 : endMarkerOffset - endLabelHalf;
+
+      if (startLabelOffset + endLabelOffset < maxPadding) {
+        if (
+          this.state.paddingLeft !== startLabelOffset ||
+          this.state.paddingRight !== endLabelOffset
+        ) {
```

Comment:
Here we are on 3rd nested if level, it is hard to read. Please try and refactor it a bit.